File: src/code-monkey/src/tools/create_file_tool.py
import os
from typing import Dict, Any
from .io_tool import IOTool
from .utils import make_file_path

class CreateFileTool(IOTool):
    name = "create_file"
    description = "Create a new file with optional content"
    input_schema = {
        "type": "object",
        "properties": {
            "fname": {
                "type": "string",
                "description": "Name of the file to create.",
            },
            "content": {
                "type": "string",
                "description": "Initial content of the file (optional).",
            },
        },
        "required": ["fname"],
    }

    def handle_tool_call(self, input: Dict[str, Any]) -> Dict[str, Any] | None:
        name = input["fname"]
        content = input.get("content", "")
        file_path = make_file_path(name)

        if os.path.exists(file_path):
            raise FileExistsError(f"The file {name} already exists.")

        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        with open(file_path, "w") as file:
            file.write(content)

        self.track_modified_file(file_path)

File: code-monkey-ts2/tools/create_file_tool.ts
import * as fs from "fs";
import * as path from "path";
import { z } from "zod";
import { instrument, currentSpan } from "../instrumentation";
import { IOTool } from "./io_tool";
import { CodeContext } from "../code_context";

const schema = z.object({
  fname: z.string().describe("Name of the file to create."),
  content: z
    .string()
    .optional()
    .describe("Initial content of the file (optional)."),
});

export class CreateFileTool extends IOTool {
  name = "create_file";
  description = "Create a new file with optional content";
  schema = schema;

  constructor(codeContext: CodeContext) {
    super(codeContext);
  }

  @instrument("Tool._call", { tool: "CreateFileTool" })
  async _call({ fname, content }: z.infer<typeof schema>): Promise<string> {
    currentSpan().setAttributes({
      fname,
      content: content || "",
    });
    try {
      const filePath = this.codeContext.resolveFile(fname);
      fs.mkdirSync(path.dirname(filePath), { recursive: true });
      fs.writeFileSync(filePath, content || "");
      await this.notifyFileModified(fname);
      return "file successfully created";
    } catch (error) {
      console.error(`Failed to create file: ${fname}`);
      console.error(error);
      return "failed to create file";
    }
  }
}